cairo/ethereum/cancun/vm/instructions/memory_instructions.cairo

```diff
@@ -0,0 +1,259 @@
+    if (cast(err3, felt) != 0) {
+        return err3;
+    }
+
+    // OPERATION
+    let memory = evm.value.memory;
+    with memory {
+        let value_bytes = memory_read_bytes(start_position, U256(new U256Struct(32, 0)));
+        expand_by(extend_memory.value.expand_by);
+    }
+    let value = Helpers.bytes32_to_uint256(value_bytes.value.data);
+    with stack {
+        let err3 = push(U256(new U256Struct(value.low, value.high)));
+        if (cast(err3, felt) != 0) {
+            return err3;
+        }
+    }
+
+    // PROGRAM COUNTER
+    EvmImpl.set_pc_stack_memory(Uint(evm.value.pc.value + 1), stack, memory);
+    let ok = cast(0, ExceptionalHalt*);
+    return ok;
+}
+
+// @notice Push the size of active memory in bytes onto the stack
+func msize{range_check_ptr, evm: Evm}() -> ExceptionalHalt* {
+    alloc_locals;
+    // STACK
+
+    // GAS
+    let err1 = charge_gas(Uint(GasConstants.GAS_BASE));
+    if (cast(err1, felt) != 0) {
+        return err1;
+    }
+
+    // OPERATION
+    let stack = evm.value.stack;
+    with stack {
+        let err2 = push(U256(new U256Struct(evm.value.memory.value.len, 0)));
+        if (cast(err2, felt) != 0) {
+            return err2;
+        }
+    }
+
+    // PROGRAM COUNTER
+    EvmImpl.set_pc_stack(Uint(evm.value.pc.value + 1), stack);
+    let ok = cast(0, ExceptionalHalt*);
+    return ok;
+}
+
+// @notice Copy the bytes in memory from one location to another
+func mcopy{range_check_ptr, evm: Evm}() -> ExceptionalHalt* {
+    alloc_locals;
+    // STACK
+    let stack = evm.value.stack;
+    with stack {
+        let (destination, err1) = pop();
+        if (cast(err1, felt) != 0) {
+            return err1;
+        }
+        let (source, err2) = pop();
+        if (cast(err2, felt) != 0) {
+            return err2;
+        }
+        let (length, err3) = pop();
+        if (cast(err3, felt) != 0) {
+            return err3;
+        }
+    }
+
+    // GAS
+    // OutOfGasError if length.high > 2**128
+    if (length.value.high != 0) {
+        tempvar err = new ExceptionalHalt(OutOfGasError);
+        return err;
+    }
+
+    let length_ceil32 = ceil32(Uint(length.value.low));
+    let (words, _) = divmod(length_ceil32.value, 32);
+    let copy_gas_cost = GasConstants.GAS_COPY * words;
+
+    let (mem_access_tuples: TupleU256U256Struct*) = alloc();
+    assert mem_access_tuples[0] = TupleU256U256Struct(
+        source, U256(new U256Struct(length.value.low, 0))
+    );
+    assert mem_access_tuples[1] = TupleU256U256Struct(
+        destination, U256(new U256Struct(length.value.low, 0))
+    );
+    tempvar mem_access_list = ListTupleU256U256(
+        new ListTupleU256U256Struct(new TupleU256U256(mem_access_tuples), 2)
+    );
+    let extend_memory = calculate_gas_extend_memory(evm.value.memory, mem_access_list);
+
+    // copy_gas_cost in [0, 3 * 2**120)
+    // extend_memory.value.cost.value in [0, 2**110)
+    // -> sum < felt_size, no overflow
```

Comment:
should we add invariants? such as `assert sum < PRIME - 1` (this doesnt work since it may overflow and if sum wraps around, then obv it'll be less than prime but still, maybe use overflowing math)

c.f. [primeagen](https://www.youtube.com/watch?v=jc_6ZkkOUpQ) and [tiger beetle](https://github.com/tigerbeetle/tigerbeetle/blob/main/docs/TIGER_STYLE.md)

plus it'll help us when we switch primes